### PR TITLE
Add support and recipes for HF models via AutoModelForCausalLM

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?
Can be invoked with `nemo` or via the example

```
nemo llm pretrain --factory 'HfAutoModel(model_name="meta-llama/Llama-3.2-1B", num_gpus_per_node=1)'
```

the example:
```
cd examples/llm/sft/
python3 hf.py --model=meta-llama/Llama-3.2-1B --devices=1 --strategy=ddp
```

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/gpt/data/hf_dataset.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            for key in ['tokens', 'labels']
+        }
+
+    def train_dataloader(self, collate_fn=None):
+        from nemo.lightning.data import add_megatron_sampler
+
+        if collate_fn is None:
+            collate_fn = lambda x: HfDatasetDataModule.collate_fn(x, pad_token_id=self.pad_token_id)
+
+        return add_megatron_sampler(
```
Comment (marcromeyn) [inline nemo/collections/llm/gpt/data/hf_dataset.py:60]: If these models are not intended to be trained with mcore, why do we need the megatron-sampler?

File: nemo/collections/llm/gpt/model/hf_auto_model.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        assert self._tokenizer is None
+        self._tokenizer = value
+
+    @staticmethod
+    def configure_tokenizer(model_name):
+        return AutoTokenizer(model_name)
+
+    def configure_model(self):
+        # create all your layers here
+        self.model = AutoModelForCausalLM.from_pretrained(self.model_name, torch_dtype='auto')
```
Comment (marcromeyn) [inline nemo/collections/llm/gpt/model/hf_auto_model.py:48]: If this is just for CausalLM, shouldn't that be in the name of the class?

File: nemo/collections/llm/recipes/hf_auto_model.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+from nemo import lightning as nl
+from nemo.collections import llm
+from nemo.collections.llm.api import finetune, pretrain
+from nemo.collections.llm.gpt.data.mock import MockDataModule
+from nemo.collections.llm.gpt.model.hf_auto_model import HfAutoModel
+from nemo.collections.llm.recipes.log.default import default_log, default_resume, tensorboard_logger
+from nemo.collections.llm.recipes.optim.adam import pytorch_adam_with_cosine_annealing
+from nemo.utils.exp_manager import TimingCallback
+
+NAME = "HfAutoModel"
```
Comment (marcromeyn) [inline nemo/collections/llm/recipes/hf_auto_model.py:32]: All other recipes use snake-case, let's do the same here.

File: examples/llm/sft/hf.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
+import fiddle as fdl
+import pytorch_lightning as pl
+from pytorch_lightning.loggers import WandbLogger
+from torch.utils.data import DataLoader
+
+from nemo import lightning as nl
+from nemo.collections import llm
+
+
+class SquadDataModuleWithMbs(llm.SquadDataModule):
```
[the remark this answers is not in the document]
Comment (akoumpa) [reply examples/llm/sft/hf.py:10]: testing

File: nemo/collections/llm/gpt/data/hf_dataset.py (hunk 1 above)
Comment (akoumpa) [reply nemo/collections/llm/gpt/data/hf_dataset.py:60]: I want to make data and sampler a plug & play thing, so if you happen to have organized your dataset with mcore that should not prevent you from using a HF model.

File: nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        assert self._tokenizer is None
+        self._tokenizer = value
+
+    @staticmethod
+    def configure_tokenizer(model_name):
+        return AutoTokenizer(model_name)
+
+    def configure_model(self):
+        # create all your layers here
+        self.model = AutoModelForCausalLM.from_pretrained(self.model_name, torch_dtype='auto')
```
[the remark this answers is not in the document]
Comment (akoumpa) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py]: Thanks, added an option for this, please LMK what you think.

File: nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        self.model.train()
+
+    def forward(self, input_ids, attention_mask=None, labels=None, loss_mask=None):
```
[the remark this answers is not in the document]
Comment (akoumpa) [reply nemo/collections/llm/gpt/model/hf_auto_model_for_causal_lm.py:58]: https://github.com/NVIDIA/NeMo/pull/10962/files#diff-78f81f4094cfea056c177e87c0d527b9ce27cee11813138e5a2a69370b922c19R270 yes, thank you.

File: nemo/lightning/data.py [hunk 7]
```diff
@@ -172,18 +174,15 @@ def add_megatron_sampler(
...
         batch_sampler = MegatronPretrainingSampler(
             total_samples=len(dataloader.dataset),
             consumed_samples=consumed_samples,
             micro_batch_size=micro_batch_size,
             global_batch_size=global_batch_size,
             rampup_batch_size=rampup_batch_size,
-            data_parallel_rank=parallel_state.get_data_parallel_rank(),
-            data_parallel_size=parallel_state.get_data_parallel_world_size(),
+            data_parallel_rank=rank,
+            data_parallel_size=world_size,
```
Comment (hemildesai) [inline nemo/lightning/data.py:185]: Why do we need these changes? It feels risky to pass in rank and world_size as args instead of infer from parallel_state?

Comment (hemildesai) [reply nemo/lightning/data.py:185]: Can we initialize them inside this method instead of making them an arg?

Comment (akoumpa) [reply nemo/lightning/data.py:185]: I want to be able to use this from the HF model and that does not use parallel_state, so I pass them as params.

Comment (hemildesai) [reply nemo/lightning/data.py:185]: In that case, can you use https://pytorch.org/docs/stable/distributed.html#torch.distributed.get_rank for HF model?

Comment (akoumpa) [reply nemo/lightning/data.py:185]: for mcore, I'm going to [use](https://github.com/NVIDIA/NeMo/pull/10962/files#diff-c423cc7981d5621754bb3e5e509fceaa9adb8537b269b7def042995cfa21d529R74) `parallel_state` since it already has that information and for hf I'm going to use torch.distributed.get_rank/get_world_size as in hf_dataset.py

File: nemo/collections/llm/gpt/data/hf_dataset.py (hunk 1 above)
Comment (akoumpa) [reply nemo/collections/llm/gpt/data/hf_dataset.py:82]: I made this optional, the use case is hf-dataset <> mcore-model.

**Review by hemildesai — APPROVED** [review]
LGTM

**Review by hemildesai — APPROVED** [review]
